drv/lpc55-sprot-server/src/main.rs

```diff
@@ -223,117 +234,181 @@ impl Io {
-        // CSn de-asserted.
         let mut bytes_received = 0;
         let mut rx = rx_buf.iter_mut();
-        while !self.spi.ssd() {
+
+        // Check for the SOT bit on first fifo read to see if we are
+        // synchronized.
+        let mut first_read = true;
+
+        let mut read_fifo = || {
             while self.spi.has_entry() {
                 bytes_received += 2;
-                let read = self.spi.read_u16();
+                let (read, sot) = self.spi.read_u16_with_sot();
+                if first_read {
+                    first_read = false;
+                    if !sot {
+                        self.stats.desynchronized =
+                            self.stats.desynchronized.wrapping_add(1);
+                        return Err(IoError::Desynchronized);
+                    }
+                }
                 let upper = (read >> 8) as u8;
                 let lower = read as u8;
                 rx.next().map(|b| *b = upper);
                 rx.next().map(|b| *b = lower);
             }
-        }
+            Ok(())
+        };
 
-        self.spi.ssd_clear();
+        // Go into a tight loop receiving as many bytes as we can until we see
+        // CSn de-asserted.
+        //
+        // This is the realtime part of the code
+        while !self.spi.ssd() {
+            read_fifo()?;
+        }
 
         // There may be bytes left in the rx fifo after CSn is de-asserted
         while self.spi.has_entry() {
-            bytes_received += 2;
-            let read = self.spi.read_u16();
-            let upper = (read >> 8) as u8;
-            let lower = read as u8;
-            rx.next().map(|b| *b = upper);
-            rx.next().map(|b| *b = lower);
+            read_fifo()?;
         }
 
+        ringbuf_entry!(Trace::ReceivedBytes(bytes_received));
+
         self.check_for_rx_error()?;
 
+        // Was this a CSn pulse?
         if bytes_received == 0 {
-            // This was a CSn pulse
             self.stats.csn_pulses = self.stats.csn_pulses.wrapping_add(1);
             return Err(IoError::Flush);
         }
 
-        ringbuf_entry!(Trace::ReceivedBytes(bytes_received));
+        Ok(bytes_received)
+    }
 
-        // We don't bother sending bytes when receiving. So we must clear
-        // the underrun error condition before we handle a reply.
+    pub fn cleanup(&mut self) {
+        // Drain our TX and RX fifos
+        self.spi.drain();
+
+        // Clear any errors
+        self.spi.rxerr_clear();
         self.spi.txerr_clear();
 
-        Ok(bytes_received)
+        // Ensure our SSA/SSD bits are not set
+        self.spi.ssa_clear();
+        self.spi.ssd_clear();
     }
 
-    fn reply(&mut self, tx_buf: &[u8]) {
+    pub fn cleanup_after_request(&mut self) -> Result<(), DesynchronizedError> {
+        let result = self.wait_for_csn_deasserted();
+        self.cleanup();
+        result
+    }
+
+    // Reply to the SP
+    // Our fifos are guaranteed to be empty at this point
```

Comment:
Is there any way we could add an assert/error for this?

Reply:
Not really. There's also always an inherent TOCTUO here. We could check the fifos, but they may not be empty right after the check.

If they aren't empty we'll end up with some sort of error here or at the SP. I'll be more detailed in the comment.